File: programmers/박소민/lv1/노란불신호등.js
/*
설계
순서.
최소공배수 구한다음에 그 때까지 시간 증가시키면서 돌리기
그냥 단순하게 초로 계산
*/

function solution(signals) {
  const n = signals.length;

  const sumSignals = signals.map(([g, y, r]) => g + y + r);

  const gcd = (a, b) => (b === 0 ? a : gcd(b, a % b));
  const lcm = (a, b) => (a * b) / gcd(a, b);

  const totalLcm = sumSignals.reduce((acc, val) => lcm(acc, val));

  for (let t = 1; t <= totalLcm; t++) {
    let isYellow = true;
    for (let i = 0; i < n; i++) {
      const [g, y, r] = signals[i];
      const sumSignal = sumSignals[i];
      const curr = (t - 1) % sumSignal;

      if (curr >= g && curr < g + y) continue;

      isYellow = false;
    }

    if (isYellow) return t;
  }

  return -1;
}

File: programmers/박소민/lv1/노란불신호등.py
def gcd(a, b):
    if b == 0:
        return a

    return gcd(b, a % b)
    
def lcm(a, b):
    return int((a * b) / gcd(a, b))

def solution(signals):
    n = len(signals)
    
    total_signals = [0] * n
    
    for i in range(n):
        [g, y, r] = signals[i]
        
        total_signals[i] = g + y + r
        
    total_lcm = 1
    
    for val in total_signals:
        total_lcm = lcm(total_lcm, val)
    
    for t in range(1, total_lcm + 1):
        is_yellow = True
        
        for i in range(n):
            [g, y, r] = signals[i]
            curr = (t - 1) % total_signals[i]
            
            if g > curr or curr >= g + y:
                is_yellow = False
                break
        
        if is_yellow:
            return t
            
    return -1